Common/Orders/Fills/IFillModel.cs

```diff
@@ -13,61 +13,799 @@
+        private bool _checkedMarketOnCloseFill;
+        private Func<Security, MarketOrder, OrderEvent> _marketFill;
+        private Func<Security, StopMarketOrder, OrderEvent> _stopMarketFill;
+        private Func<Security, StopLimitOrder, OrderEvent> _stopLimitFill;
+        private Func<Security, LimitOrder, OrderEvent> _limitFill;
+        private Func<Security, MarketOnOpenOrder, OrderEvent> _marketOnOpenFill;
+        private Func<Security, MarketOnCloseOrder, OrderEvent> _marketOnCloseFill;
+
+        /// <summary>
+        /// The <see cref="SubscriptionDataConfig"/> provider to use
+        /// </summary>
+        protected ISubscriptionDataConfigProvider SubscriptionDataConfigProvider;
+
+        /// <summary>
+        /// Initializes and creates a new instance
+        /// </summary>
+        protected IFillModel()
+        {
+            _self = Expression.Constant(this);
+            _securityParameter = Expression.Parameter(typeof(Security), "security");
+        }
+
+        /// <summary>
+        /// Return an order event with the fill details
+        /// </summary>
+        /// <param name="context">A context object containing the security and order</param>
+        /// <returns>Order fill information detailing the average price and quantity filled.</returns>
+        public virtual OrderEvent Fill(FillModelContext context)
+        {
+            var order = context.Order;
+            SubscriptionDataConfigProvider = context.ConfigProvider;
+
+            switch (order.Type)
+            {
+                case OrderType.Market:
+                    return MarketFill(context);
+                case OrderType.Limit:
+                    return LimitFill(context);
+                case OrderType.StopMarket:
+                    return StopMarketFill(context);
+                case OrderType.StopLimit:
+                    return StopLimitFill(context);
+                case OrderType.MarketOnOpen:
+                    return MarketOnOpenFill(context);
+                case OrderType.MarketOnClose:
+                    return MarketOnCloseFill(context);
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+        }
+
         /// <summary>
         /// Model the slippage on a market order: fixed percentage of order price
         /// </summary>
-        /// <param name="asset">Asset we're trading this order</param>
-        /// <param name="order">Order to update</param>
+        /// <param name="context">A context object containing the security and order</param>
         /// <returns>Order fill information detailing the average price and quantity filled.</returns>
-        OrderEvent MarketFill(Security asset, MarketOrder order);
+        protected virtual OrderEvent MarketFill(FillModelContext context)
+        {
+            if (!_checkedMarketFill)
+            {
+                _checkedMarketFill = true;
+                SetFunctionCall(context.Order.Type);
```

Comment:
this might be less confusing in this method if it's set up as `_marketFill = CreateFunctionCall(context.Order.Type)` vs having the method set the member and then using it a few lines later. This is a 'hidden assignment' from this context. The function, can be a _pure_ function and it will improve understandability here.